pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,325 @@
+						return s.Name()
+					}), ","))
+			}
+			q.Remove(hash, cmd)
+			continue
+		}
+		q.RateLimitingInterface.Done(hash)
+		// Requeue command if not complete
+		q.RateLimitingInterface.AddRateLimited(hash)
+	}
+	logging.FromContext(ctx).Errorf("deprovisioning queue is broken and has shutdown")
+}
+
+// Add adds commands to the Queue
+// Each command added to the queue should already be validated and ready for execution.
+func (q *Queue) Add(cmd *Command) error {
+	providerIDs := lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+		return s.ProviderID()
+	})
+	// First check if we can add the command.
+	if err := q.CanAdd(providerIDs...); err != nil {
+		return fmt.Errorf("adding command, %w", err)
+	}
+	hash, err := Hash(cmd)
+	if err != nil {
+		return err
+	}
+	for _, candidate := range cmd.Candidates {
+		q.candidateProviderIDToCommandID[candidate.ProviderID()] = hash
+		q.idToCommand[hash] = cmd
+	}
+	// Idempotently mark for deletion
+	q.cluster.MarkForDeletion(providerIDs...)
+	q.RateLimitingInterface.Add(hash)
+	return nil
+}
+
+// Remove fully clears the queue of all references of a hash/command
+func (q *Queue) Remove(hash uint64, cmd *Command) {
+	q.RateLimitingInterface.Forget(hash)
+	for _, candidate := range cmd.Candidates {
+		delete(q.candidateProviderIDToCommandID, candidate.ProviderID())
+	}
+	delete(q.idToCommand, hash)
+	q.RateLimitingInterface.Done(hash)
+}
+
+// CanAdd is a quick check to see if the candidate is already part of a deprovisioning action
+func (q *Queue) CanAdd(ids ...string) error {
+	var err error
+	for _, id := range ids {
+		if _, ok := q.candidateProviderIDToCommandID[id]; ok {
+			err = multierr.Append(err, fmt.Errorf("candidate is part of active deprovisioning decision"))
+		}
+	}
+	return err
+}
+
+// Handle will check for timeouts, then execute a wait or termination.
+// If there
+func (q *Queue) Handle(ctx context.Context, cmd *Command) (bool, error) {
+	// logging.FromContext(ctx).Infof("DEBUGGING: This is the current time: %s, this is the time added: %s", q.clock)
+	if q.clock.Since(cmd.TimeAdded) > maxRetryDuration {
+		return false, fmt.Errorf("abandoning command, reached timeout, %w", cmd.LastError)
+	}
+	// If the time hasn't expired, either wait, or terminate.
+	requeue, err := q.WaitOrTerminate(ctx, cmd)
+	if err != nil {
+		// If there was an error, set this as the command's last error so that we can propagate it.
+		cmd.LastError = err
+		if !retry.IsRecoverable(err) {
+			return false, nil
+		}
+		return requeue, nil
+	}
+	return requeue, nil
+}
+
+// WaitOrTerminate will wait until launched machines are ready.
+// Once the replacements are ready, it will terminate the candidates.
+// Will return true if the item in the queue should be re-queued. If a command has
+// timed out, this will return false.
+// nolint:gocyclo
+func (q *Queue) WaitOrTerminate(ctx context.Context, cmd *Command) (bool, error) {
```

Comment:
Is it somehow possible for us to collapse the (bool, error) return into just an error. Is there a case where we don't return an error for failure but need a requeue?